File: static/js/uf.js
// ==========================================
// GLOBAL STATE
// ==========================================

window.currentDevice = null;
window.currentParameter = null;
window.currentInterval = "30min";
window.currentStart = "";
window.currentEnd = "";


// ==========================================
// DEVICE BUTTON CLICK
// ==========================================

function selectDevice(parameter, device) {

    // If clicking same device again → TOGGLE OFF
    if (window.currentDevice === device &&
        window.currentParameter === parameter) {

        clearGraph();
        window.currentDevice = null;
        window.currentParameter = null;
        return;
    }

    // Otherwise activate
    window.currentDevice = device;
    window.currentParameter = parameter;

    loadData();
}


// ==========================================
// INTERVAL BUTTON CLICK
// ==========================================

function setInterval(interval) {

    window.currentInterval = interval;

    // Only reload if device is selected
    if (window.currentDevice && window.currentParameter) {
        loadData();
    }
}


// ==========================================
// CUSTOM TIME RANGE
// ==========================================

function loadCustomRange() {

    const start = document.getElementById("startTime").value;
    const end = document.getElementById("endTime").value;

    if (!start || !end) {
        alert("Please select both start and end time.");
        return;
    }

    window.currentStart = start;
    window.currentEnd = end;

    if (window.currentDevice && window.currentParameter) {
        loadData();
    }
}


// ==========================================
// LOAD DATA FROM BACKEND
// ==========================================

function loadData() {

    fetch(`/api/uf-data?parameter=${encodeURIComponent(window.currentParameter)}`
        + `&device=${encodeURIComponent(window.currentDevice)}`
        + `&interval=${window.currentInterval}`
        + `&start=${window.currentStart}`
        + `&end=${window.currentEnd}`)

    .then(response => response.json())
    .then(data => {

        if (!data.timestamps || data.timestamps.length === 0) {
            clearGraph();
            return;
        }

        updateGraph(data.timestamps, data.values);
        updateTable(data);
    })
    .catch(error => {
        console.error("UF FETCH ERROR:", error);
        clearGraph();
    });
}


// ==========================================
// CLEAR GRAPH + TABLE
// ==========================================

function clearGraph() {

    if (window.ufChart) {
        window.ufChart.destroy();
    }

    document.getElementById("latestValue").innerText = "-";
    document.getElementById("meanValue").innerText = "-";
    document.getElementById("minValue").innerText = "-";
    document.getElementById("maxValue").innerText = "-";
    document.getElementById("dataTime").innerText = "-";
}


// ==========================================
// UPDATE GRAPH
// ==========================================

function updateGraph(labels, values) {

    const ctx = document.getElementById("ufChart").getContext("2d");

    if (window.ufChart) {
        window.ufChart.destroy();
    }

    window.ufChart = new Chart(ctx, {
        type: "line",
        data: {
            labels: labels,
            datasets: [{
                label: window.currentParameter,
                data: values,
                fill: false,
                tension: 0.1
            }]
        },
        options: {
            responsive: true,
            maintainAspectRatio: false
        }
    });
}


// ==========================================
// UPDATE KEY-VALUE TABLE
// ==========================================

function updateTable(data) {

    document.getElementById("latestValue").innerText = data.latest;
    document.getElementById("meanValue").innerText = data.mean;
    document.getElementById("minValue").innerText = data.min;
    document.getElementById("maxValue").innerText = data.max;
    document.getElementById("dataTime").innerText = data.data_time;
}

File: routes/uf.py
from flask import Blueprint, render_template, jsonify, request
from database.uf_data import get_uf_device_data

uf_bp = Blueprint("uf", __name__)


# ==============================
# UF PAGE
# ==============================
@uf_bp.route("/uf")
def uf_page():
    return render_template("uf_system.html")


# ==============================
# UF API
# ==============================
@uf_bp.route("/api/uf-data")
def uf_data_api():

    parameter = request.args.get("parameter")
    device = request.args.get("device")
    interval = request.args.get("interval", "30min")

    # 🔥 ADD THESE TWO LINES
    start = request.args.get("start")
    end = request.args.get("end")

    if not parameter or not device:
        return jsonify({
            "timestamps": [],
            "values": [],
            "latest": 0,
            "mean": 0,
            "min": 0,
            "max": 0,
            "data_time": "-"
        })

    try:
        # 🔥 PASS start and end to backend function
        data = get_uf_device_data(parameter, device, interval, start, end)
        return jsonify(data)

    except Exception as e:
        print("UF API ERROR:", e)

        return jsonify({
            "timestamps": [],
            "values": [],
            "latest": 0,
            "mean": 0,
            "min": 0,
            "max": 0,
            "data_time": "-"
        })